File: src/bindings-python/tests/golden_test.py
"""Golden-file validation, mirroring the wasm and JVM suites.

Every template in the corpus is validated through both engines at both detail
levels, and the result must match resources/expected/validation_reports.json exactly
(up to the fields the golden file intentionally excludes). The typed uniffi
records are serialized back into serde's JSON shape (camelCase names, enum
names, unwrapped JsonValue variants, Nones omitted), so this also proves the
Python type surface is faithful to the serialized report shape.
"""

import copy
import enum
import json
import os
import re
import unittest

from cloudformation_validate import (
    CelEngine,
    EntityType,
    JsonValue,
    RegoEngine,
    Severity,
    ValidateConfig,
)

TESTS_DIR = os.path.dirname(os.path.abspath(__file__))
WORKSPACE = os.path.dirname(os.path.dirname(TESTS_DIR))
TEMPLATES_ROOT = os.path.join(WORKSPACE, "resources", "templates")
GOLDEN_FILE = os.path.join(WORKSPACE, "resources", "expected", "validation_reports.json")

GOLDEN_DIRS = ["bad", "cdk", "good", "gh-issues", "integration", "issues", "lsp", "public", "quickstart"]

# Fields present only in detailed reports; stripped from the golden entry when
# comparing standard reports.
DETAILED_ONLY_DIAGNOSTIC_FIELDS = ["documentationUrl", "context", "ruleDescription", "phase", "section"]

_CAMEL = re.compile(r"_([a-z0-9])")


def _camel_case(name):
    return _CAMEL.sub(lambda m: m.group(1).upper(), name)


def _unwrap_json_value(value):
    if isinstance(value, JsonValue.NULL):
        return None
    if isinstance(value, (JsonValue.BOOL, JsonValue.INT, JsonValue.FLOAT, JsonValue.STRING)):
        return value.value
    if isinstance(value, JsonValue.ARRAY):
        return [_unwrap_json_value(item) for item in value.items]
    if isinstance(value, JsonValue.OBJECT):
        return {key: _unwrap_json_value(entry) for key, entry in value.entries.items()}
    raise AssertionError(f"unhandled JsonValue variant: {value!r}")


def to_jsonable(obj):
    """Converts a uniffi record tree into serde's serialized JSON shape."""
    if obj is None or isinstance(obj, (bool, int, float, str)):
        return obj
    if isinstance(obj, EntityType):
        # Serde serializes EntityType variants verbatim (PascalCase), unlike
        # the SCREAMING_SNAKE forms of the other report enums.
        return "".join(word.capitalize() for word in obj.name.split("_"))
    if isinstance(obj, enum.Enum):
        return obj.name
    if isinstance(obj, JsonValue):
        return _unwrap_json_value(obj)
    if isinstance(obj, list):
        return [to_jsonable(item) for item in obj]
    if isinstance(obj, dict):
        return {key: to_jsonable(value) for key, value in obj.items()}
    return {
        _camel_case(field): to_jsonable(value) for field, value in vars(obj).items() if value is not None
    }


def discover_golden_templates():
    templates = []
    for sub in GOLDEN_DIRS:
        root = os.path.join(TEMPLATES_ROOT, sub)
        if not os.path.isdir(root):
            continue
        for dirpath, _, filenames in os.walk(root):
            for filename in filenames:
                if filename.endswith((".yaml", ".yml", ".json")):
                    full = os.path.join(dirpath, filename)
                    templates.append(os.path.relpath(full, TEMPLATES_ROOT).replace(os.sep, "/"))
    return sorted(templates)


def strip_golden_excluded_fields(report, file_path=None):
    if file_path is not None:
        report["filePath"] = file_path
    report.pop("version", None)
    report.pop("performance", None)
    if isinstance(report.get("metadata"), dict):
        report["metadata"].pop("rulesEvaluated", None)
        report["metadata"].pop("cfnLintVersion", None)
        report["metadata"].pop("resourceSchemaVersion", None)
    return report


def strip_detailed_only_fields(report):
    for diagnostic in report.get("diagnostics", []):
        for field in DETAILED_ONLY_DIAGNOSTIC_FIELDS:
            diagnostic.pop(field, None)
    return report


with open(GOLDEN_FILE, encoding="utf-8") as f:
    GOLDEN = json.load(f)

EXPECTED_TEMPLATES = discover_golden_templates()

DEBUG_LEVEL = ValidateConfig(severity_level=Severity.DEBUG)

REGO = RegoEngine()
CEL = CelEngine()


class GoldenFileValidationTest(unittest.TestCase):
    maxDiff = None

    def setUp(self):
        self.assertTrue(EXPECTED_TEMPLATES, "no templates discovered")

    def assert_matches_golden(self, engine, detailed):
        for rel in EXPECTED_TEMPLATES:
            with self.subTest(template=rel):
                self.assertIn(rel, GOLDEN, f"{rel}: missing golden entry")
                path = os.path.join(TEMPLATES_ROOT, rel)
                if detailed:
                    report = engine.validate_detailed(path, DEBUG_LEVEL)
                    expected = strip_golden_excluded_fields(copy.deepcopy(GOLDEN[rel]))
                else:
                    report = engine.validate_standard(path, DEBUG_LEVEL)
                    expected = strip_detailed_only_fields(strip_golden_excluded_fields(copy.deepcopy(GOLDEN[rel])))
                actual = strip_golden_excluded_fields(to_jsonable(report), rel)
                self.assertEqual(expected, actual, f"{rel}: report does not match golden")

    def test_rego_detailed_matches_golden(self):
        self.assert_matches_golden(REGO, detailed=True)

    def test_rego_standard_matches_golden(self):
        self.assert_matches_golden(REGO, detailed=False)

    def test_cel_detailed_matches_golden(self):
        self.assert_matches_golden(CEL, detailed=True)

    def test_cel_standard_matches_golden(self):
        self.assert_matches_golden(CEL, detailed=False)


class PerformanceMetricsTest(unittest.TestCase):
    def test_performance_present_with_timing_per_phase(self):
        report = REGO.validate_detailed(os.path.join(TEMPLATES_ROOT, "good", "generic.yaml"), DEBUG_LEVEL)
        performance = report.performance
        for phase in (
            "schema_init",
            "engine_init",
            "model_build",
            "schema_validate",
            "rule_evaluation",
            "diagnostic_finalize",
            "validate_total",
        ):
            metric = getattr(performance, phase)
            self.assertIsInstance(metric.duration_ms, float, f"performance.{phase}.duration_ms")
            self.assertGreaterEqual(metric.duration_ms, 0.0)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: src/bindings-go/go/golden_test.go
// Golden-file validation, mirroring the wasm and JVM suites: every template in
// the corpus is validated through both engines at both detail levels, and the
// result must match resources/expected/all_templates.json exactly (up to the
// fields the golden file intentionally excludes). Reports round-trip through
// the typed Go structs before comparison, so this also proves the Go type
// surface is faithful to the serialized report shape.
package cfnvalidate_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"

	cfnvalidate "github.com/aws-cloudformation/cloudformation-validate/src/bindings-go/go"
)

var goldenDirs = []string{"bad", "cdk", "good", "gh-issues", "integration", "issues", "lsp", "public", "quickstart"}

// Fields present only in detailed reports; stripped from the golden entry when
// comparing standard reports.
var detailedOnlyDiagnosticFields = []string{"documentationUrl", "context", "ruleDescription", "phase", "section"}

var templatesRoot = filepath.Join(workspaceDir, "resources", "templates")

func discoverGoldenTemplates(t *testing.T) []string {
	t.Helper()
	var templates []string
	for _, sub := range goldenDirs {
		root := filepath.Join(templatesRoot, sub)
		if _, err := os.Stat(root); err != nil {
			continue
		}
		err := filepath.WalkDir(root, func(path string, entry os.DirEntry, err error) error {
			if err != nil {
				return err
			}
			if entry.IsDir() {
				return nil
			}
			switch filepath.Ext(entry.Name()) {
			case ".yaml", ".yml", ".json":
				rel, relErr := filepath.Rel(templatesRoot, path)
				if relErr != nil {
					return relErr
				}
				templates = append(templates, filepath.ToSlash(rel))
			}
			return nil
		})
		if err != nil {
			t.Fatalf("discovering templates under %s: %v", root, err)
		}
	}
	sort.Strings(templates)
	return templates
}

func loadGolden(t *testing.T) map[string]map[string]any {
	t.Helper()
	content, err := os.ReadFile(filepath.Join(workspaceDir, "resources", "expected", "all_templates.json"))
	if err != nil {
		t.Fatalf("reading golden file: %v", err)
	}
	var golden map[string]map[string]any
	if err := json.Unmarshal(content, &golden); err != nil {
		t.Fatalf("decoding golden file: %v", err)
	}
	return golden
}

// stripGoldenExcludedFields removes the report fields the golden file excludes
// (version, performance, and changing metadata provenance) and normalizes filePath.
func stripGoldenExcludedFields(report map[string]any, filePath string) map[string]any {
	if filePath != "" {
		report["filePath"] = filePath
	}
	delete(report, "version")
	delete(report, "performance")
	if metadata, ok := report["metadata"].(map[string]any); ok {
		delete(metadata, "rulesEvaluated")
		delete(metadata, "cfnLintVersion")
		delete(metadata, "resourceSchemaVersion")
	}
	return report
}

func stripDetailedOnlyFields(report map[string]any) map[string]any {
	diagnostics, ok := report["diagnostics"].([]any)
	if !ok {
		return report
	}
	for _, entry := range diagnostics {
		if d, ok := entry.(map[string]any); ok {
			for _, field := range detailedOnlyDiagnosticFields {
				delete(d, field)
			}
		}
	}
	return report
}

// toComparable round-trips a typed report through JSON into the same generic
// shape as the golden entries.
func toComparable(t *testing.T, report any) map[string]any {
	t.Helper()
	data, err := json.Marshal(report)
	if err != nil {
		t.Fatalf("marshaling report: %v", err)
	}
	var out map[string]any
	if err := json.Unmarshal(data, &out); err != nil {
		t.Fatalf("re-decoding report: %v", err)
	}
	return out
}

func cloneGoldenEntry(t *testing.T, entry map[string]any) map[string]any {
	t.Helper()
	data, err := json.Marshal(entry)
	if err != nil {
		t.Fatalf("cloning golden entry: %v", err)
	}
	var out map[string]any
	if err := json.Unmarshal(data, &out); err != nil {
		t.Fatalf("cloning golden entry: %v", err)
	}
	return out
}

func diffJSON(t *testing.T, rel string, actual, expected map[string]any) {
	t.Helper()
	if reflect.DeepEqual(actual, expected) {
		return
	}
	actualJSON, _ := json.MarshalIndent(actual, "", " ")
	expectedJSON, _ := json.MarshalIndent(expected, "", " ")
	t.Errorf("%s: report does not match golden\n--- actual ---\n%s\n--- expected ---\n%s", rel, actualJSON, expectedJSON)
}

func TestGoldenFileValidation(t *testing.T) {
	templates := discoverGoldenTemplates(t)
	if len(templates) == 0 {
		t.Fatal("no templates discovered")
	}
	golden := loadGolden(t)
	debugLevel := &cfnvalidate.ValidateConfig{SeverityLevel: cfnvalidate.SeverityDebug}

	for engineName, engine := range bothEngines(t) {
		t.Run(engineName+" detailed matches golden", func(t *testing.T) {
			for _, rel := range templates {
				expected, ok := golden[rel]
				if !ok {
					t.Errorf("%s: missing golden entry", rel)
					continue
				}
				report, err := engine.ValidateDetailedFile(filepath.Join(templatesRoot, rel), debugLevel)
				if err != nil {
					t.Errorf("%s: validation failed: %v", rel, err)
					continue
				}
				actual := stripGoldenExcludedFields(toComparable(t, report), rel)
				want := stripGoldenExcludedFields(cloneGoldenEntry(t, expected), "")
				diffJSON(t, rel, actual, want)
			}
		})

		t.Run(engineName+" standard matches golden", func(t *testing.T) {
			for _, rel := range templates {
				expected, ok := golden[rel]
				if !ok {
					t.Errorf("%s: missing golden entry", rel)
					continue
				}
				report, err := engine.ValidateStandardFile(filepath.Join(templatesRoot, rel), debugLevel)
				if err != nil {
					t.Errorf("%s: validation failed: %v", rel, err)
					continue
				}
				actual := stripGoldenExcludedFields(toComparable(t, report), rel)
				want := stripDetailedOnlyFields(stripGoldenExcludedFields(cloneGoldenEntry(t, expected), ""))
				diffJSON(t, rel, actual, want)
			}
		})
	}
}

func TestPerformanceMetricsPresent(t *testing.T) {
	engine := mustEngine(t, cfnvalidate.NewRegoEngine, nil)
	report, err := engine.ValidateDetailedFile(filepath.Join(templatesRoot, "good", "generic.yaml"), nil)
	if err != nil {
		t.Fatalf("validation failed: %v", err)
	}
	phases := map[string]cfnvalidate.PhaseMetric{
		"schemaInit":         report.Performance.SchemaInit,
		"engineInit":         report.Performance.EngineInit,
		"modelBuild":         report.Performance.ModelBuild,
		"schemaValidate":     report.Performance.SchemaValidate,
		"ruleEvaluation":     report.Performance.RuleEvaluation,
		"diagnosticFinalize": report.Performance.DiagnosticFinalize,
		"validateTotal":      report.Performance.ValidateTotal,
	}
	for name, metric := range phases {
		if metric.DurationMs < 0 {
			t.Errorf("performance.%s.durationMs = %v, want >= 0", name, metric.DurationMs)
		}
	}
	if report.Performance.ValidateTotal.DurationMs == 0 {
		t.Error("validateTotal duration must be recorded")
	}
}

func TestEmptyTemplateReportsFatalParseRule(t *testing.T) {
	for name, engine := range bothEngines(t) {
		report, err := engine.ValidateStandardFile(filepath.Join(templatesRoot, "empty.yaml"), nil)
		if err != nil {
			t.Fatalf("%s: validation failed: %v", name, err)
		}
		if report.Status != cfnvalidate.StatusError {
			t.Errorf("%s: status = %s, want ERROR", name, report.Status)
		}
		if len(report.Diagnostics) == 0 || report.Diagnostics[0].RuleID != "F1101" ||
			report.Diagnostics[0].Severity != cfnvalidate.SeverityFatal {
			t.Errorf("%s: first diagnostic must be FATAL F1101, got %+v", name, report.Diagnostics)
		}
	}
}

func TestTemplateModelSections(t *testing.T) {
	minimal, err := os.ReadFile(filepath.Join(templatesRoot, "good", "minimal.yaml"))
	if err != nil {
		t.Fatalf("reading minimal.yaml: %v", err)
	}
	model, err := cfnvalidate.ParseTemplate(minimal)
	if err != nil {
		t.Fatalf("parse failed: %v", err)
	}
	defer model.Destroy()
	if v := model.FormatVersion(); v == nil || *v != "2010-09-09" {
		t.Errorf("formatVersion = %v, want 2010-09-09", v)
	}
	resourcesJSON, err := model.Resources()
	if err != nil {
		t.Fatalf("Resources failed: %v", err)
	}
	var resources map[string]any
	if err := json.Unmarshal(resourcesJSON, &resources); err != nil {
		t.Fatalf("decoding resources: %v", err)
	}
	if _, ok := resources["IamPipeline"]; !ok {
		t.Error("minimal.yaml must contain IamPipeline resource")
	}
	if transforms, _ := model.Transforms(); len(transforms) != 0 {
		t.Errorf("transforms = %v, want none", transforms)
	}

	generic, err := os.ReadFile(filepath.Join(templatesRoot, "good", "generic.yaml"))
	if err != nil {
		t.Fatalf("reading generic.yaml: %v", err)
	}
	genericModel, err := cfnvalidate.ParseTemplate(generic)
	if err != nil {
		t.Fatalf("parse failed: %v", err)
	}
	defer genericModel.Destroy()
	if d := genericModel.Description(); d == nil || *d != "A sample template" {
		t.Errorf("description = %v, want 'A sample template'", d)
	}
	conditions, err := genericModel.Conditions()
	if err != nil {
		t.Fatalf("Conditions failed: %v", err)
	}
	if !containsString(conditions, "ProdVolumeSize") {
		t.Errorf("conditions = %v, want ProdVolumeSize", conditions)
	}
	outputsJSON, err := genericModel.Outputs()
	if err != nil {
		t.Fatalf("Outputs failed: %v", err)
	}
	var outputs map[string]any
	if err := json.Unmarshal(outputsJSON, &outputs); err != nil {
		t.Fatalf("decoding outputs: %v", err)
	}
	if _, ok := outputs["ElasticIP"]; !ok {
		t.Error("generic.yaml outputs must contain ElasticIP")
	}
	diagnosticJSON, err := genericModel.DiagnosticModel()
	if err != nil {
		t.Fatalf("DiagnosticModel failed: %v", err)
	}
	var diagnostic map[string]any
	if err := json.Unmarshal(diagnosticJSON, &diagnostic); err != nil {
		t.Fatalf("decoding diagnostic model: %v", err)
	}
	for _, section := range []string{"template", "resources"} {
		if _, ok := diagnostic[section]; !ok {
			t.Errorf("diagnostic model must contain %q section", section)
		}
	}

	malformed, err := os.ReadFile(filepath.Join(templatesRoot, "malformed.yaml"))
	if err != nil {
		t.Fatalf("reading malformed.yaml: %v", err)
	}
	if _, err := cfnvalidate.ParseTemplate(malformed); err == nil {
		t.Error("malformed YAML must fail to parse")
	}
}

func containsString(values []string, want string) bool {
	for _, v := range values {
		if v == want {
			return true
		}
	}
	return false
}

func TestCombinedCustomAndGuardRuleListings(t *testing.T) {
	celConfig := &cfnvalidate.EngineConfig{
		CustomRules: []cfnvalidate.ExternalRuleSource{{Name: "cel_multi_custom.json", Content: loadRule(t, "cel_multi_custom.json")}},
		GuardRules: []cfnvalidate.ExternalRuleSource{
			{Name: "guard_encryption.guard", Content: loadRule(t, "guard_encryption.guard")},
			{Name: "guard_multi.guard", Content: loadRule(t, "guard_multi.guard")},
		},
	}
	regoConfig := &cfnvalidate.EngineConfig{
		CustomRules: []cfnvalidate.ExternalRuleSource{{Name: "rego_multi_custom.rego", Content: loadRule(t, "rego_multi_custom.rego")}},
		GuardRules: []cfnvalidate.ExternalRuleSource{
			{Name: "guard_encryption.guard", Content: loadRule(t, "guard_encryption.guard")},
			{Name: "guard_multi.guard", Content: loadRule(t, "guard_multi.guard")},
		},
	}
	cel := mustEngine(t, cfnvalidate.NewCelEngine, celConfig)
	rego := mustEngine(t, cfnvalidate.NewRegoEngine, regoConfig)

	// Rego discovers custom rule metadata during evaluation.
	if _, err := rego.ValidateStandardFile(filepath.Join(templatesRoot, "bad", "invalid_deletion_policy.yaml"), nil); err != nil {
		t.Fatalf("rego warm-up validation failed: %v", err)
	}

	type expectedRule struct {
		severity    cfnvalidate.Severity
		origin      cfnvalidate.RuleOrigin
		description string
	}
	expected := map[string]expectedRule{
		"CUSTOM010":               {cfnvalidate.SeverityError, cfnvalidate.RuleOriginCustom, "S3 bucket must have versioning enabled"},
		"CUSTOM011":               {cfnvalidate.SeverityWarn, cfnvalidate.RuleOriginCustom, "S3 bucket should have lifecycle rules configured"},
		"check_bucket_encryption": {"", cfnvalidate.RuleOriginGuard, "S3 bucket must have encryption configured"},
		"check_bucket_versioning": {"", cfnvalidate.RuleOriginGuard, "S3 bucket must have versioning enabled"},
		"check_bucket_lifecycle":  {"", cfnvalidate.RuleOriginGuard, "S3 bucket should have lifecycle rules configured"},
	}

	lists := map[string][]cfnvalidate.RuleInfo{}
	for name, engine := range map[string]*cfnvalidate.Engine{"cel": cel, "rego": rego} {
		rules, err := engine.ListRules()
		if err != nil {
			t.Fatalf("%s: ListRules failed: %v", name, err)
		}
		if !sort.SliceIsSorted(rules, func(i, j int) bool { return rules[i].ID < rules[j].ID }) {
			t.Errorf("%s: rules must be sorted by id", name)
		}
		byID := map[string]cfnvalidate.RuleInfo{}
		for _, r := range rules {
			byID[r.ID] = r
		}
		for id, want := range expected {
			rule, ok := byID[id]
			if !ok {
				t.Errorf("%s: rule %s must be listed", name, id)
				continue
			}
			if rule.Origin != want.origin {
				t.Errorf("%s: %s origin = %s, want %s", name, id, rule.Origin, want.origin)
			}
			if rule.Description != want.description {
				t.Errorf("%s: %s description = %q, want %q", name, id, rule.Description, want.description)
			}
			if want.severity != "" && rule.Severity != want.severity {
				t.Errorf("%s: %s severity = %s, want %s", name, id, rule.Severity, want.severity)
			}
		}
		lists[name] = rules
	}

	celJSON, _ := json.Marshal(lists["cel"])
	regoJSON, _ := json.Marshal(lists["rego"])
	if !strings.EqualFold(string(celJSON), string(regoJSON)) {
		t.Error("CEL and Rego must list identical rules with custom + guard sources")
	}
}
